### PR TITLE
feat: update logic for new notification creation

### PR DESCRIPTION
**Pull request by kshitij-k-osmosys**
### Pre-requisites

- [x] I have gone through the Contributing guidelines for [Submitting a Pull Request (PR)](../../CONTRIBUTING.md#submitting-a-pull-request-pr) and ensured that this is not a duplicate PR.
- [ ] I have performed preliminary testing using the [test suite](../../apps/api/OsmoX.postman_collection.json) to ensure that any existing features are not impacted and any new features are working as expected.
- [x] I have updated the required [api docs](../../apps/api/docs/) as applicable.
- [ ] I have added/updated test cases to the [test suite](../../apps/api/OsmoX.postman_collection.json) as applicable

### PR Details

PR details have been updated as per the given format (see below)

- [x] PR title adheres to the format specified in guidelines (e.g., `feat: add admin login endpoint`)
- [x] Description has been added
- [x] Related changes have been added (optional)
- [x] Screenshots have been added (optional)
- [x] Query request and response examples have been added (as applicable, in case added or updated)
- [x] Documentation changes have been listed (as applicable)
- [ ] Test suite output is added (as applicable)
- [ ] Pending actions have been added (optional)
- [ ] Any other additional notes have been added (optional)

### Additional Information

- [x] Appropriate label(s) have been added (`ready for review` should be added if the PR is ready to be reviewed)
- [x] Assignee(s) and reviewer(s) have been added (optional)

---

**Description:**

Updated the new notification creation logic to add application Id based on the server api key used.

**Related changes:**

- Create new `server-api-key` entity with relevant service functions.
- Create new `application` entity with relevant service functions.
- Update the `new notification` creation logic.
- Now ApplicationId and CreatedBy, UpdatedBy will be assigned based on key used.
- Use the updated migration format

**Screenshots:**

api-success2

**Query request and response:**

#### cURL

```sh
curl --location 'http://localhost:3000/notifications' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer mysecuretoken' \
--data-raw '{
  "channelType": 2,
  "data": {
    "from": "osmo@foundation.com",
    "to": "test@email.com",
    "subject": "Osmo subject",
    "text": "This is a test notification",
    "html": "<b>This is a test notification</b>"
  }
}'
```

#### Response

```json
{
    "status": "success",
    "data": {
        "notification": {
            "channelType": 2,
            "data": {
                "from": "osmo@foundation.com",
                "to": "test@email.com",
                "subject": "Osmo subject",
                "text": "This is a test notification",
                "html": "<b>This is a test notification</b>"
            },
            "applicationId": 1001,
            "createdBy": "sampleOsmoXApp",
            "updatedBy": "sampleOsmoXApp",
            "result": null,
            "id": 8,
            "deliveryStatus": 1,
            "createdOn": "2024-04-15T09:51:07.000Z",
            "updatedOn": "2024-04-15T09:51:07.000Z",
            "status": 1
        }
    }
}
```

**Documentation changes:**

- Add new output for `applicationId`, `createdBy`, `updatedBy`

**Test suite output:**

- Add the output of the status of different test cases in the testing suite.

**Pending actions:**

- Add list of any pending actions that have or would require to be done in this PR.

**Additional notes:**

- Add list of any additional notes you may want to convey in this PR.

## Summary by CodeRabbit

- **New Features**
  - Enhanced API key management with new columns and constraints in the database.
  - Introduced modules for managing applications and server API keys with corresponding services and resolvers.
  - Added GraphQL types for creating and updating applications and server API keys.

- **Bug Fixes**
  - Modified notification creation method to include authentication via request headers.

- **Documentation**
  - Added comprehensive test suites for newly introduced services and resolvers.

- **Refactor**
  - Updated database schema and seeded sample data for improved functionality.

### CODE REVIEW

File: apps/api/src/modules/notifications/notifications.controller.ts [hunk 1]
```diff
@@ -25,11 +26,17 @@ export class NotificationsController {
   @Post()
   @UseGuards(ApiKeyGuard)
   async addNotification(
+    @Req() request: Request,
     @Body() notificationData: CreateNotificationDto,
   ): Promise<Record<string, unknown>> {
     try {
-      const createdNotification =
-        await this.notificationService.createNotification(notificationData);
+      const authHeader = request.headers['authorization'];
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/notifications/notifications.controller.ts:33]: ApiKeyGuard is doing the validation

File: apps/api/src/modules/notifications/notifications.service.ts [hunk 2]
```diff
@@ -36,11 +43,67 @@ export class NotificationsService {
...
+
+    // Set correct application name using applicationId
+    const applicationEntry = await this.applicationsResolver.findApplicationById(
+      apiKeyEntry.applicationId,
+    );
+
+    if (!applicationEntry) {
+      throw new Error('Related Application does not exist');
+    }
+
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/notifications/notifications.service.ts]: Added in new commit

File: apps/api/src/modules/notifications/notifications.service.ts [hunk 3]
```diff
@@ -36,11 +43,60 @@ export class NotificationsService {
       throw new BadRequestException(`Channel ${notification.channelType} is not enabled`);
     }
 
-    notification.createdBy = this.configService.getOrThrow<string>('APP_NAME') || 'OsmoX';
-    notification.updatedBy = this.configService.getOrThrow<string>('APP_NAME') || 'OsmoX';
+    // Set correct ApplicationId
+    notification.applicationId = await this.getApplicationIdFromApiKey(authHeader);
+
+    // Set correct application name using applicationId
+    notification.createdBy = 'OsmoX';
```
Comment (LakshayaT) [inline apps/api/src/modules/notifications/notifications.service.ts:50]: Created By value should also be fetched from the application ID only.

File: apps/api/src/database/migrations/1692870736646-seeddata.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/database/migrations/1692870736646-seeddata.ts:1]: Will remove this

File: apps/api/src/modules/applications/dto/create-application.input.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { InputType, Int, Field } from '@nestjs/graphql';
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/applications/dto/create-application.input.ts:1]: These were auto generated
Will remove all dtos for now

File: apps/api/src/modules/applications/dto/update-application.input.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import { CreateApplicationInput } from './create-application.input';
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/applications/dto/update-application.input.ts:1]: These were auto generated
Will remove all dtos for now

File: apps/api/src/modules/server-api-keys/dto/create-server-api-key.input.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import { InputType, Int, Field } from '@nestjs/graphql';
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/server-api-keys/dto/create-server-api-key.input.ts:1]: These were auto generated
Will remove all dtos for now

File: apps/api/src/modules/server-api-keys/dto/update-server-api-key.input.ts [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import { CreateServerApiKeyInput } from './create-server-api-key.input';
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/server-api-keys/dto/update-server-api-key.input.ts:1]: These were auto generated
Will remove all dtos for now

File: apps/api/src/modules/notifications/notifications.service.ts (hunk 3 above)
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/notifications/notifications.service.ts:50]: Ok

File: apps/api/src/modules/notifications/notifications.service.ts [hunk 9]
```diff
@@ -36,11 +43,60 @@ export class NotificationsService {
...
+    } catch (error) {
+      this.logger.log('Error creating notification:', error.message);
+      throw error;
+    }
+  }
+
+  // Get correct application name using applicationId
+  async getApplicationNameFromId(applicationId: number): Promise<string> {
+    try {
+      const applicationEntry = await this.applicationsResolver.findApplicationById(applicationId);
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/notifications/notifications.service.ts]: Using services instead

File: apps/api/src/modules/notifications/notifications.service.ts [hunk 10]
```diff
@@ -36,11 +43,60 @@ export class NotificationsService {
...
+        apiKeyToken = bearerToken.substring(7);
+      } else {
+        throw new Error('Invalid bearer token format');
+      }
+
+      if (apiKeyToken == null) {
+        throw new Error('Failed to assign applicationId');
+      }
+
+      const apiKeyEntry = await this.serverApiKeysResolver.findApiKey(apiKeyToken);
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/notifications/notifications.service.ts]: Using services instead

File: apps/api/src/modules/applications/applications.resolver.ts [hunk 11]
```diff
@@ -0,0 +1,15 @@
+import { Resolver, Query, Args } from '@nestjs/graphql';
+import { ApplicationsService } from './applications.service';
+import { Application } from './entities/application.entity';
+
+@Resolver(() => Application)
+export class ApplicationsResolver {
```
Comment (LakshayaT) [inline apps/api/src/modules/applications/applications.resolver.ts:6]: We shouldn't have an unprotected resolver for this. I also think we do not need this for the time being. If possible, we can remove this resolver entirely for now (along with the spec file) and add it later when needed.

File: apps/api/src/modules/server-api-keys/server-api-keys.resolver.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
+import { Resolver, Query, Args } from '@nestjs/graphql';
+import { ServerApiKeysService } from './server-api-keys.service';
+import { ServerApiKey } from './entities/server-api-key.entity';
+
+@Resolver(() => ServerApiKey)
+export class ServerApiKeysResolver {
```
Comment (LakshayaT) [inline apps/api/src/modules/server-api-keys/server-api-keys.resolver.ts:6]: We shouldn't have an unprotected resolver for this. I also think we do not need this for the time being. If possible, we can remove this resolver entirely for now (along with the spec file) and add it later when we need.

**Review by LakshayaT — CHANGES_REQUESTED** [review]
@kshitij-k-osmosys Please have a look at these comments as well.

File: apps/api/src/modules/server-api-keys/server-api-keys.resolver.ts (hunk 12 above)
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/server-api-keys/server-api-keys.resolver.ts]: Removed this

File: apps/api/src/modules/applications/applications.resolver.ts (hunk 11 above)
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/applications/applications.resolver.ts]: Removed this

**Review by LakshayaT — APPROVED** [review]

**Pull request merged**